cmd/immuadmin/command/hot_backup.go

```diff
@@ -0,0 +1,661 @@
+
+	var bar *progressbar.ProgressBar
+	if progress {
+		bar = progressbar.New(-1)
+	}
+
+	done := make(chan struct{}, 1)
+	defer close(done)
+	terminated := make(chan os.Signal, 1)
+	signal.Notify(terminated, os.Interrupt)
+
+	stop := false
+	go func() {
+		select {
+		case <-done:
+		case <-terminated:
+			fmt.Fprintf(os.Stderr, "Terminated by signal\n")
+			stop = true
+		}
+	}()
+
+	if firstTx != 0 {
+		// first transaction was already read during DB verification, so process it first
+		err := cl.restoreTx(signature, payload)
+		if err != nil {
+			return err
+		}
+		startTx = firstTx
+		lastTx = firstTx
+	}
+
+	for !stop {
+		tx, signature, payload, err := getTx(input)
+		if errors.Is(err, io.EOF) {
+			break
+		}
+		if err != nil {
+			return err
+		}
+
+		err = cl.restoreTx(signature, payload)
+		if err != nil {
+			return err
+		}
+
+		if startTx == 0 {
+			startTx = tx
+		}
+		lastTx = tx
+		if bar != nil {
+			bar.Add(1)
+		}
+	}
+	if startTx == 0 {
+		fmt.Printf("Target database is up-to-date, nothing restored\n")
+	} else {
+		fmt.Printf("Restored transactions %d - %d\n", startTx, lastTx)
+	}
+
+	return nil
+}
+
+func (cl *commandlineHotBck) restoreTx(signature, payload []byte) error {
+	maxPayload := uint32(cl.options.MaxRecvMsgSize)
+
+	stream, err := cl.immuClient.ReplicateTx(cl.context)
+	if err != nil {
+		return err
+	}
+	offset := uint32(0)
+	remainder := uint32(len(payload))
+	for remainder > 0 {
+		chunkSize := maxPayload
+		if remainder < maxPayload {
+			chunkSize = remainder
+		}
+		err = stream.Send(&schema.Chunk{Content: payload[offset : offset+chunkSize]})
+		if err != nil {
+			return err
+		}
+		remainder -= chunkSize
+		offset += chunkSize
+	}
+	metadata, err := stream.CloseAndRecv()
+	if err != nil {
+		return err
+	}
+	if !bytes.Equal(signature, metadata.EH) {
+		return fmt.Errorf("transaction signatures don't match")
+	}
+
+	return nil
+}
+
+// lastTxInFile assumes that file is positioned on transaction boundary
+// it also can be used to validate file correctness
+// in case of success file pointer is positioned to file end
+func lastTxInFile(file io.Reader) (uint64, []byte, error) {
+	last := uint64(0)
+	var signature []byte
+	for {
+		var cur uint64
+		var err error
+		cur, signature, _, err = getTx(file)
+		if errors.Is(err, io.EOF) {
+			break
+		}
+		if err != nil {
+			return 0, nil, err
+		}
+
+		if cur != last+1 && last > 0 {
+			return 0, nil, ErrTxWrongOrder
+		}
+		last = cur
+	}
+
+	return last, signature, nil
+}
+
+// returns tx smallest possible ID which is greater of equal to the one requested
+func findTx(file io.Reader, tx uint64) (txID uint64, signature []byte, payload []byte, err error) {
+	for {
+		txID, signature, payload, err = getTx(file)
+		if err != nil {
+			return
+		}
+		if txID >= tx {
+			break
+		}
+	}
+
+	return
+}
+
+// read transaction, position stream pointer just after the transaction
+func getTx(file io.Reader) (txId uint64, signature []byte, tx []byte, err error) {
```

Comment:
readTx / nextTx, so to transmit the notion of a cursor as it's not an idempotent function

Reply:
Agree, will change

Reply:
changed